File: examples/vision/matting/rvm/python/infer.py
import fastdeploy as fd
import cv2
import os


def parse_arguments():
    import argparse
    import ast
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--model", required=True, help="Path of RobustVideoMatting model.")
    parser.add_argument("--image", type=str, help="Path of test image file.")
    parser.add_argument("--video", type=str, help="Path of test video file.")
    parser.add_argument(
        "--bg",
        type=str,
        required=True,
        default=None,
        help="Path of test background image file.")
    parser.add_argument(
        '--output-composition',
        type=str,
        default="composition.mp4",
        help="Path of composition video file.")
    parser.add_argument(
        '--output-alpha',
        type=str,
        default="alpha.mp4",
        help="Path of alpha video file.")
    parser.add_argument(
        "--device",
        type=str,
        default='cpu',
        help="Type of inference device, support 'cpu' or 'gpu'.")
    parser.add_argument(
        "--use_trt",
        type=ast.literal_eval,
        default=False,
        help="Wether to use tensorrt.")
    return parser.parse_args()


def build_option(args):
    option = fd.RuntimeOption()
    if args.device.lower() == "gpu":
        option.use_gpu()
    if args.use_trt:
        option.use_trt_backend()
        option.set_trt_input_shape("src", [1, 3, 1920, 1080])
        option.set_trt_input_shape("r1i", [1, 1, 1, 1], [1, 16, 240, 135],
                                   [1, 16, 240, 135])
        option.set_trt_input_shape("r2i", [1, 1, 1, 1], [1, 20, 120, 68],
                                   [1, 20, 120, 68])
        option.set_trt_input_shape("r3i", [1, 1, 1, 1], [1, 40, 60, 34],
                                   [1, 40, 60, 34])
        option.set_trt_input_shape("r4i", [1, 1, 1, 1], [1, 64, 30, 17],
                                   [1, 64, 30, 17])

    return option


args = parse_arguments()
output_composition = args.output_composition
output_alpha = args.output_alpha

# 配置runtime，加载模型
runtime_option = build_option(args)
model = fd.vision.matting.RobustVideoMatting(
    args.model, runtime_option=runtime_option)
bg = cv2.imread(args.bg)

if args.video is not None:
    # for video
    cap = cv2.VideoCapture(args.video)
    # Define the codec and create VideoWriter object
    fourcc = cv2.VideoWriter_fourcc(*'mp4v')
    composition = cv2.VideoWriter(output_composition, fourcc, 20.0,
                                  (1080, 1920))
    alpha = cv2.VideoWriter(output_alpha, fourcc, 20.0, (1080, 1920))

    frame_id = 0
    while True:
        frame_id = frame_id + 1
        _, frame = cap.read()
        if frame is None:
            break
        result = model.predict(frame)
        vis_im = fd.vision.vis_matting(frame, result)
        vis_im_with_bg = fd.vision.swap_background_matting(frame, bg, result)
        alpha.write(vis_im)
        composition.write(vis_im_with_bg)
        cv2.waitKey(30)
    cap.release()
    composition.release()
    alpha.release()
    cv2.destroyAllWindows()
    print("Visualized result video save in {} and {}".format(
        output_composition, output_alpha))

if args.image is not None:
    # for image
    im = cv2.imread(args.image)
    result = model.predict(im.copy())
    print(result)
    # 可视化结果
    vis_im = fd.vision.vis_matting(im, result)
    vis_im_with_bg = fd.vision.swap_background_matting(im, bg, result)
    cv2.imwrite("visualized_result_fg.jpg", vis_im)
    cv2.imwrite("visualized_result_replaced_bg.jpg", vis_im_with_bg)
    print(
        "Visualized result save in ./visualized_result_replaced_bg.jpg and ./visualized_result_fg.jpg"
    )

File: examples/vision/matting/rvm/cpp/infer.cc
// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "fastdeploy/vision.h"

#ifdef WIN32
const char sep = '\\';
#else
const char sep = '/';
#endif

void CpuInfer(const std::string& model_file, const std::string& image_file,
              const std::string& background_file) {
  auto option = fastdeploy::RuntimeOption();
  auto model = fastdeploy::vision::matting::RobustVideoMatting(model_file, "", option);
  if (!model.Initialized()) {
    std::cerr << "Failed to initialize." << std::endl;
    return;
  }

  auto im = cv::imread(image_file);
  auto im_bak = im.clone();
  cv::Mat bg = cv::imread(background_file);
  fastdeploy::vision::MattingResult res;
  if (!model.Predict(&im, &res)) {
    std::cerr << "Failed to predict." << std::endl;
    return;
  }
  auto vis_im = fastdeploy::vision::VisMatting(im_bak, res);
  auto vis_im_with_bg =
      fastdeploy::vision::SwapBackground(im_bak, bg, res);
  cv::imwrite("visualized_result.jpg", vis_im_with_bg);
  cv::imwrite("visualized_result_fg.png", vis_im);
  std::cout << "Visualized result save in ./visualized_result.jpg "
               "and ./visualized_result_fg.jpg"
            << std::endl;
}

void GpuInfer(const std::string& model_file, const std::string& image_file,
              const std::string& background_file) {
  auto option = fastdeploy::RuntimeOption();
  option.UseGpu();
  auto model = fastdeploy::vision::matting::RobustVideoMatting(model_file, "", option);
  if (!model.Initialized()) {
    std::cerr << "Failed to initialize." << std::endl;
    return;
  }

  auto im = cv::imread(image_file);
  auto im_bak = im.clone();
  cv::Mat bg = cv::imread(background_file);
  fastdeploy::vision::MattingResult res;
  if (!model.Predict(&im, &res)) {
    std::cerr << "Failed to predict." << std::endl;
    return;
  }
  auto vis_im = fastdeploy::vision::VisMatting(im_bak, res);
  auto vis_im_with_bg =
      fastdeploy::vision::SwapBackground(im_bak, bg, res);
  cv::imwrite("visualized_result.jpg", vis_im_with_bg);
  cv::imwrite("visualized_result_fg.png", vis_im);
  std::cout << "Visualized result save in ./visualized_result_replaced_bg.jpg "
               "and ./visualized_result_fg.jpg"
            << std::endl;
}

void TrtInfer(const std::string& model_file, const std::string& image_file,
              const std::string& background_file) {
  auto option = fastdeploy::RuntimeOption();
  option.UseGpu();
  option.UseTrtBackend();
  option.SetTrtInputShape("src", {1, 3, 1920, 1080});
  option.SetTrtInputShape("r1i", {1, 1, 1, 1}, {1, 16, 240, 135}, {1, 16, 240, 135});
  option.SetTrtInputShape("r2i", {1, 1, 1, 1}, {1, 20, 120, 68}, {1, 20, 120, 68});
  option.SetTrtInputShape("r3i", {1, 1, 1, 1}, {1, 40, 60, 34}, {1, 40, 60, 34});
  option.SetTrtInputShape("r4i", {1, 1, 1, 1}, {1, 64, 30, 17}, {1, 64, 30, 17});
  auto model = fastdeploy::vision::matting::RobustVideoMatting(model_file, "", option);
  if (!model.Initialized()) {
    std::cerr << "Failed to initialize." << std::endl;
    return;
  }

  auto im = cv::imread(image_file);
  auto im_bak = im.clone();
  cv::Mat bg = cv::imread(background_file);
  fastdeploy::vision::MattingResult res;
  if (!model.Predict(&im, &res)) {
    std::cerr << "Failed to predict." << std::endl;
    return;
  }
  auto vis_im = fastdeploy::vision::VisMatting(im_bak, res);
  auto vis_im_with_bg =
      fastdeploy::vision::SwapBackground(im_bak, bg, res);
  cv::imwrite("visualized_result.jpg", vis_im_with_bg);
  cv::imwrite("visualized_result_fg.png", vis_im);
  std::cout << "Visualized result save in ./visualized_result.jpg "
               "and ./visualized_result_fg.jpg"
            << std::endl;
}

int main(int argc, char* argv[]) {
  if (argc < 5) {
    std::cout
        << "Usage: infer_demo path/to/model_dir path/to/image run_option, "
           "e.g ./infer_model ./rvm_mobilenetv3_fp32.onnx ./test.jpg ./test_bg.jpg 0"
        << std::endl;
    std::cout << "The data type of run_option is int, 0: run with cpu; 1: run "
                 "with gpu; 2: run with gpu and use tensorrt backend."
              << std::endl;
    return -1;
  }
  if (std::atoi(argv[4]) == 0) {
    CpuInfer(argv[1], argv[2], argv[3]);
  } else if (std::atoi(argv[4]) == 1) {
    GpuInfer(argv[1], argv[2], argv[3]);
  } else if (std::atoi(argv[4]) == 2) {
    TrtInfer(argv[1], argv[2], argv[3]);
  }
  return 0;
}
